### ISSUE TITLE
Optimize isA calls when the type is a JS type and not a user interop type

### ISSUE THREAD

**Issue by srujzs** [dart-lang/sdk#56130]
Currently, calls like `isA<JSArray>()` are lowered to a call to [`instanceOfString`](https://github.com/dart-lang/sdk/blob/3961a27bd71f1bf4b370e60ab05a9c3389ac0512/sdk/lib/js_interop/js_interop.dart#L371) in `dart:js_interop`. While correct, this method does additional checking of parts to make sure the fully qualified name of an interop type is available. With JS types, they're available in the global context and therefore don't need that additional checking. Instead, we should have a private helper that the lowering uses that gets the constructor and does an `instanceof` without needing to do any part checking.

cc @mattrberry

**Comment by srujzs** [comment]
Yup! We use kernel transformers to accomplish this. See [`_createIsACheck`](https://github.com/dart-lang/sdk/blob/e41b531a59f458b520db70e145fdc9b3bcd8ecb7/pkg/_js_interop_checks/lib/src/transformations/shared_interop_transformer.dart#L416) for more details.

> Instead, we should have a private helper that the lowering uses that gets the constructor and does an instanceof without needing to do any part checking.

To clarify what I mean by this as well:

- We use "patch" files in the SDK to patch some `external` members (like all the `toJS`) calls. These files are usually compiler-dependent. For the JS compilers, we use a shared implementation [here](https://github.com/dart-lang/sdk/blob/e41b531a59f458b520db70e145fdc9b3bcd8ecb7/sdk/lib/_internal/js_shared/lib/js_interop_patch.dart#L4) and for dart2wasm, we use [this](https://github.com/dart-lang/sdk/blob/e41b531a59f458b520db70e145fdc9b3bcd8ecb7/sdk/lib/_internal/wasm/lib/js_interop_patch.dart#L4). The methods that get "patched" with the `@patch` annotation are replaced with their patched implementation in the kernel.
- These patch files are in the same library as the library they're patching e.g. `dart:js_interop`, but hidden from users. Therefore, we can add a private method in here that the kernel transformer can use to do a "direct" `instanceof` check and avoid the string parsing of `instanceOfString`.

**Comment by rutvik110** [comment]
gotcha! Thanx for the added info. Will be helpful.

**Comment by rutvik110** [comment]
@srujzs How can one debug changes made in any of the sdk packages inside `pkg`. For example, I want to debug some changes I made in `shared_interop_transformer` but  not sure how to approach it. I believe I would've to regenerate the dart sdk for the kernel transformer changes to be picked up by the compiler? but doing that for every change doesn't seem optimal. So likely I'm missing something or is that the right way to go?

**Comment by srujzs** [comment]
Since transformers and web compilers are actually run by the Dart VM, it's possible to debug that code directly and iirc, you don't need to build the SDK every time you make a change (that's only for runtime files and such). I use VS Code and set up a configuration using . For example, my configuration to debug `pkg/dart2wasm` looks like this:

```dart
        {
            "name": "dart2wasm compile",
            "program": "<path_to_sdk>/pkg/dart2wasm/bin/dart2wasm.dart",
            "args": [
                "--platform=<path_to_sdk>/xcodebuild/ReleaseARM64/dart2wasm_platform.dill",
                "<path_to_test_file>",
                "out.wasm"
            ],
            "request": "launch",
            "type": "dart",
        },
```

I believe you can debug the other web compilers using `pkg/dev_compiler/bin/dartdevc.dart` and `pkg/compiler/lib/src/dart2js.dart` as the 'program' values.

**Comment by rutvik110** [comment]
ok so I was able to get the dart2js compiler running. Btw for the `js/static_interop_test/js_types_test.dart` the compilation seems to be failing due to an unrecognized `pragma` annotation. Removing that seems to compile the test file successfully.

```bash
Warning: The 'dart2js' entrypoint script is deprecated, please use 'dart compile js' instead.

file:///Users/rutviktak/flutter_projects/os_contributions/dart-sdk/sdk/tests/lib/js/static_interop_test/js_types_test.dart:152:1:
Error: Unrecognized dart2js pragma @pragma('dart2js:assumeDynamic')
confuse(x) => x;
^
Error: Compilation failed.

Exited (1).
```

Is this expected? or is it a potential bug in the test file?

Also, while trying out the same with dart2wasm compiler with wasm test file, got following errors, 

```bash
../wasm_builder/lib/src/ir/instructions.dart:43:58: Error: Operator '[]' cannot be called on 'Map<Instruction, StackTrace>?' because it is potentially null.
 - 'Map' is from 'dart:core'.
 - 'Instruction' is from 'package:wasm_builder/src/ir/instructions.dart' ('../wasm_builder/lib/src/ir/instructions.dart').
 - 'StackTrace' is from 'dart:core'.
      if (_stackTraces != null) s.debugTrace(_stackTraces[i]!);
                                                         ^
../wasm_builder/lib/src/builder/types.dart:130:23: Error: This requires the experimental 'records' language feature to be enabled.
Try passing the '--enable-experiment=records' command line option.
        LinkedHashMap<(int, int, List<ir.DefType>), List<List<ir.DefType>>>(
                      ^
../wasm_builder/lib/src/builder/types.dart:144:11: Error: The 'records' language feature is disabled for this library.
Try removing the package language version or setting the language version to 3.0 or higher.
          (group.length, structType.fields.length, group), () => []).add(group);
```

For the second error where it says to enable the experimental flag for records, I couldn't find an option for that in the programs inputs.

**Comment by rutvik110** [comment]
@srujzs Regarding your suggested solution for this issue, 
` Instead, we should have a private helper that the lowering uses that gets the constructor and does an instanceof without needing to do any part checking.`. 

Why can't we directly lower the `isA` call to `instanceof` call instead of `instanceOfString`? Are there limitations that prevent one from doing something like this?
```dart
// inside _createIsACheck method
check = StaticInvocation(_instanceof,
          Arguments(//pass object constructor));
```

Beside that, I'm looking into how I can call a private helper declared in patch files within the kernel transformer during the replacement phase but coudn't find a way to do that. Are there any examples of this I can look into?

**Comment by srujzs** [comment]
Sorry for the late response!

> Is this expected? or is it a potential bug in the test file?

I believe these pragmas are allowlisted so they are only allowed when test mode is enabled. You might need to run these with `test.py` to make them work.

> Also, while trying out the same with dart2wasm compiler with wasm test file, got following errors,

I'm not too sure what's going on there or what the right commands here are, but I'd suggest the same wrt using `test.py` e.g. `test.py -r chrome -c dart2wasm ...` instead.

> Why can't we directly lower the isA call to instanceof call instead of instanceOfString? Are there limitations that prevent one from doing something like this?

We definitely could and that's a totally fine approach (and maybe better since we don't need extra members). One benefit of having a private helper for this is that we can leverage the `JS` foreign function and avoid the round-trip of getting the constructor to then immediately do an `instanceof`. For example, a private helper may look something like (this is the JS compilers version of `JS`, a similar function is used in dart2wasm):

```dart
bool _instanceofNoRoundTrip(JSAny? o, String globalType) =>
  foreign_helper.JS('bool', '# instanceof #', o, globalType);
```

It also avoids you having to mess around with kernel nodes directly.

> Are there any examples of this I can look into?

We do something like that [here](https://github.com/dart-lang/sdk/blob/e41b531a59f458b520db70e145fdc9b3bcd8ecb7/pkg/_js_interop_checks/lib/src/transformations/js_util_optimizer.dart#L113) when transforming interop calls. We fetch the private member (which should be part of the `dart:` library) and then call it using kernel nodes later.

**Comment by rutvik110** [comment]
> We definitely could and that's a totally fine approach (and maybe better since we don't need extra members). One benefit of having a private helper for this is that we can leverage the `JS` foreign function and avoid the round-trip of getting the constructor to then immediately do an `instanceof`

oh, ok. I guess then the private helpers approach sounds good which saves us on some computations from that round-trip and directly messing with kernel nodes at cost of some extra members. I'll look into adding support for these.

**Comment by srujzs** [comment]
> Which one do you prefer btw?

I think it's okay to have an extra private member in the compilation. Ultimately, we'll be adding code anyways due to additional kernel nodes, so paying that cost once up-front by using the private helpers SGTM.

**Comment by rutvik110** [comment]
Sounds good!
